initrd/bin/kexec-save-default

```diff
@@ -33,106 +33,213 @@ TMP_MENU_FILE="/tmp/kexec/kexec_menu.txt"
 ENTRY_FILE="$paramsdir/kexec_default.$index.txt"
 HASH_FILE="$paramsdir/kexec_default_hashes.txt"
 
+
+lvm_suggest=$(lvm vgscan|awk -F '"' {'print $1'}|tail -n +2)
+num_lvm=$(echo "$lvm_suggest"|wc -l)
+if [ "$num_lvm" -eq 1 ] && [ -n "$lvm_suggest" ]; then
+  lvm_volume_group=$lvm_suggest
+elif [ -z "$lvm_suggest" ]; then
+  num_lvm=0
+fi
+
+devices_suggest=$(blkid | cut -d ':' -f 1 | while read device;do cryptsetup isLuks "$device";if [ $(echo $?) == 0 ]; then echo "$device";fi; done | sort)
+num_devices=$(echo "$devices_suggest"|wc -l)
+if [ "$num_devices" -eq 1 ]; then
+  key_devices=$devices_suggest
+else
+  devices_suggest=$(echo $devices_suggest)
+fi
+
+
 if [ ! -r "$TMP_MENU_FILE" ]; then
-	die "No menu options available, please run kexec-select-boot"
+  die "No menu options available, please run kexec-select-boot"
 fi
 
-entry=`head -n $index $TMP_MENU_FILE | tail -1`
+entry=$(head -n "$index" $TMP_MENU_FILE | tail -1)
 if [ -z "$entry" ]; then
-	die "Invalid menu index $index"
+  die "Invalid menu index $index"
 fi
 
 KEY_DEVICES="$paramsdir/kexec_key_devices.txt"
 KEY_LVM="$paramsdir/kexec_key_lvm.txt"
 save_key="n"
+
 if [ "$CONFIG_TPM" = "y" ]; then
-	if [ ! -r "$KEY_DEVICES" ]; then
-		read \
-			-n 1 \
-			-p "Do you wish to add a disk encryption to the TPM [y/N]: " \
-			add_key_confirm
-		echo
-
-		if [ "$add_key_confirm" = "y" \
-			-o "$add_key_confirm" = "Y" ]; then
-			lvm_suggest="e.g. qubes_dom0 or blank"
-			devices_suggest="e.g. /dev/sda2 or blank"
-			save_key="y"
-		fi
-	else
-		read \
-			-n 1 \
-			-p "Do you want to reseal a disk key to the TPM [y/N]: " \
-			change_key_confirm
-		echo
-
-		if [ "$change_key_confirm" = "y" \
-			-o "$change_key_confirm" = "Y" ]; then
-			old_lvm_volume_group=""
-			if [ -r "$KEY_LVM" ]; then
-				old_lvm_volume_group=`cat $KEY_LVM` || true
-				old_key_devices=`cat $KEY_DEVICES \
-				| cut -d\  -f1 \
-				| grep -v "$old_lvm_volume_group" \
-				| xargs` || true
-			else
-				old_key_devices=`cat $KEY_DEVICES \
-				| cut -d\  -f1 | xargs` || true
-			fi
-
-			lvm_suggest="was '$old_lvm_volume_group'"
-			devices_suggest="was '$old_key_devices'"
-			save_key="y"
-		fi
-	fi
-
-	if [ "$save_key" = "y" ]; then
-		echo "+++ LVM volume groups (lvm vgscan): "
-		lvm vgscan || true
-
-		read \
-			-p "Encrypted LVM group? ($lvm_suggest): " \
-			lvm_volume_group
-
-		echo "+++ Block devices (blkid): "
-		blkid || true
-
-		read \
-			-p "Encrypted devices? ($devices_suggest): " \
-			key_devices
-
-		save_key_params="-s -p $paramsdev"
-		if [ -n "$lvm_volume_group" ]; then
-			save_key_params="$save_key_params -l $lvm_volume_group $key_devices"
-		else
-			save_key_params="$save_key_params $key_devices"
-		fi
-		echo "Running kexec-save-key with params: $save_key_params"
-		kexec-save-key $save_key_params \
-			|| die "Failed to save the disk key"
-	fi
+  if [ ! -s "$KEY_DEVICES" ]; then
```

Comment:
Does this section *have* to be run when someone is saving a default boot option? I'd much prefer we move TPM LUKS keys elsewhere so this doesn't confuse users who mistakenly select "yes" here to move along with the default boot process, but end up having problems down the road when the TPM gets reset.

Can this move to a separate function dedicated to this purpose or is this something that has to be set when we save the default boot options?

Reply:
@kylerankin The code change picks up those integrity changes resulting from a rom flash and is dealt with automatically now. As said below, the user saying Y here would have to say yes to change the disk, yes to seal with TPM, and confirm that the disk partition /LVM volume is the desired one, at each step having the chance to cancel and get to the previous emnu permitting them to boot the selection only.

When does a TPM gets reset?

Reply:
The problem is that a user who *doesn't* already use the TPM for LUKS disk encryption is being prompted to set it up out of the blue when they are saving their default boot option.

*Every* user will set a default boot option, but only a few will use the TPM for LUKS encryption. If we have to prompt people in this script for this option, I'd prefer we only do it if we detect they are already using this feature. Otherwise the user should store their LUKS key in the TPM via an option in the Advanced -> TPM/TOTP/HOTP Options menu.

Reply:
@kylerankin : so the code now prompts and defaults now to setting a new default, asks the user if he wants to mofigy the disk, in which case he is asked if he wants to seal it with the TPM. If he refuses that step, I fixed it so that it just saves a default, like one would normally expect.

The point is that a Heads user SHOULD normally want to use the TPM to set a Disk Unlock Key passphrase, different then the Disk Recovery Key passphrase, which if captured, would give that person the possibility to clone the disk and read content at will, which takes only 25 minutes.

One keypoint, probably misunderstood of Heads, is to set such Disk Unlock Key, which depends on the TPM measurements and passphrase to unlock disk content, which if evesdropped, only permit an attacked to access content on the local computer.

Is that ok with you like that?

Reply:
@kylerankin : this is wy the OEM Wizard insists so much on setting the Disk Unlock Key for it's added confidentiality, privacy and security.

Reply:
If the OEM wizard does this, it's OK, I'm mostly objecting to prompting the user with this option when they are outside of the OEM wizard and are just trying to save a default boot option in kexec-save-default.

Personally my preference is to shift to allowing users to unlock their disk with their smart card instead as it's both more convenient and more secure and that's where I've been focusing my own efforts recently (even though Qubes's LUKS implementation doesn't have native support for smart card unlock, Debian does now). For users who go that route, being prompted to use the TPM for LUKS when they are just saving a default boot option could confuse them and they could say 'yes' without understanding all of the implications.

Reply:
@tlaurion If I'm reading your change correctly, it will only prompt the user if they already have kexec_key_devices.txt present. If that's the case please confirm and I'm totally fine with this change.

Reply:
@kylerankin : unfortunately no, it actually asks the user every time the default changes. It just reuses the LUKS device defined under kexec_key_devices.txt to prompt the user to reuse it and defaults to use it if the user hits enter.

@merge @ kylerankin: Is there any outside way of knowing that the GPG card is used to unlokc the LUKS container that we could check in? The actual choice of not proposing the highest security option offered by Heads because debian-unstable offers LUKS unlocking from a GPG card is not empowering the user to be able to secure its QubesOS installation the best way it is actually possible to enforce for most other Operating  Systems. 

QubesOS will support it when dom0 moves to fedora-29 if I got all the intrications right, but I would need advice onto how to properly deal with this without dropping the most secure option currently offered by Heads for something not currently supported by most Operating Systems, while QubesOS users would be directly benefiting of this enforcement.

Reply:
@kylerankin Mmmmmh... Doesn't seem to be really externally dealable with when looking at https://source.puri.sm/pureos/packages/smartcard-key-luks/blob/master/smartcard-key-luks

Reply:
@tlaurion How about you just wrap this controversial section inside of an if that gets triggered based on an environment variable/config option? You can even default to it being set to on, and I could ship firmware with it off if I wanted to and allow the user to turn it on. That way everyone can control what the default is.

Reply:
To answer your other question, there's no way from looking at the LUKS key slots to know whether someone is using a smart card to unlock it.

Reply:
@kylerankin : anything that could be looked at in the current default GRUB used option to nullify the proposition of sealing with TPM?

Reply:
Unfortunately not, it's an option that's set in cryptsetup config files in /etc/ and loaded into the initrd. You'd have to extract and inspect the initrd for Debian-specific hooks to figure out it's happening.

Reply:
In summary, LUKS has a standard secret in the key slot, cryptsetup just executes a script inside initrd that uses the smart card to decrypt a file inside the initrd that contains that secret.

Reply:
@kylerankin : modifications in https://github.com/osresearch/heads/pull/551/commits/f5271913c76bc7886fc6d7b722115c1fa6587317

Reply:
works for me.

Reply:
I'm confident you figure out something sane :) but really I think that Heads shouldn't care what the user's runtime kernel does. That's outside of the project's scope.